retis-events/src/events.rs

```diff
@@ -122,13 +122,11 @@ impl Event {
 
 impl EventFmt for Event {
     fn event_fmt(&self, f: &mut std::fmt::Formatter, format: DisplayFormat) -> std::fmt::Result {
-        // First format the first event line starting with the always-there
-        // {common} section, followed by the {kernel} or {user} one.
-        write!(
-            f,
-            "{}",
-            self.0.get(&SectionId::Common).unwrap().display(format)
-        )?;
+        // First format the first event line starting with the {common} section,
+        // followed by the {kernel} or {user} one.
+        if let Some(common) = self.0.get(&SectionId::Common) {
```

Comment:
On one side I do agree this is not common to all events, but it should be at least to most. The second thing is this is kinda part of an ABI (at least the "common" keywork in JSON), even though we could break it.

Another option: the common section has a timestamp and optional information (the smp id is not but setting 0 is perfectly fine). Maybe we actually do want it to be mandatory, so even for metadata events we have a timestamp information (eg. `Event::with_common()` to create an event for non-BPF events)?

Reply:
Hmm... zero is still a valid smp id so I don't know if we should do that. Maybe having it `Optional`?
Other than that, I think it's a good idea. I don't really like having a json object with just one field but I think it's better than having a "common" section that is not common.

Some other thought (trying to think of other pieces of information that might come handy in a "common" section): Do we need to flag an event as a metadata one? Events coming from eBPF side won't have metadata sections right? If a collector has to add data that is related to it, it can create another (normal) section.

So, maybe we need a `type=[bpf|meta]` field? or even a `meta=[true|false] # default false to avoid extra space in normal events which are more common`?

Reply:
That way we can easily identify an event as containing only metadata without having to iterate through sections.

Reply:
> Hmm... zero is still a valid smp id so I don't know if we should do that. Maybe having it `Optional`? Other than that, I think it's a good idea. I don't really like having a json object with just one field but I think it's better than having a "common" section that is not common.

Sounds good.

> Some other thought (trying to think of other pieces of information that might come handy in a "common" section): Do we need to flag an event as a metadata one? Events coming from eBPF side won't have metadata sections right? If a collector has to add data that is related to it, it can create another (normal) section.
> 
> So, maybe we need a `type=[bpf|meta]` field? or even a `meta=[true|false] # default false to avoid extra space in normal events which are more common`?

(Note that before writing this down I already made the common section to be mandatory only).

That would shift the metadata vs non-metadata to the event instead of the section level. My idea was really not to differentiate "metadata" vs other sections. However I do agree we *do* make a distinction as of now (more on that later).

I don't think an event should be metadata only or metadata free. The metadata is there to provide additional information for post-processing. It can be global information (that can be extended / amended by future events) but it could be specific to a given event.

Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.

For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.

For the event part, I think it has to bet loosely defined to get some flexibility. To me an event is a set of sections somehow related to each others. It might be to a packet event (aka. probe + timestamp) or to the initial configuration that is dumped by collectors (aka. timestamp only).

The difference in how those are consumed at post-processing time (is it global info? specific to an event?) is really only depending on the kind of section it is. That's up to the post-processing logic.

Reply:
> > Hmm... zero is still a valid smp id so I don't know if we should do that. Maybe having it `Optional`? Other than that, I think it's a good idea. I don't really like having a json object with just one field but I think it's better than having a "common" section that is not common.
> 
> Sounds good.
> 
> > Some other thought (trying to think of other pieces of information that might come handy in a "common" section): Do we need to flag an event as a metadata one? Events coming from eBPF side won't have metadata sections right? If a collector has to add data that is related to it, it can create another (normal) section.
> > So, maybe we need a `type=[bpf|meta]` field? or even a `meta=[true|false] # default false to avoid extra space in normal events which are more common`?
> 
> (Note that before writing this down I already made the common section to be mandatory only).
> 
> That would shift the metadata vs non-metadata to the event instead of the section level. My idea was really not to differentiate "metadata" vs other sections. However I do agree we _do_ make a distinction as of now (more on that later).
> 
> I don't think an event should be metadata only or metadata free. The metadata is there to provide additional information for post-processing. It can be global information (that can be extended / amended by future events) but it could be specific to a given event.
> 

> Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.
> 
> For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.
> 

I  don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.

How about we do as with `skb` sections? In commands such as `print` and `sort`, we allow the user to define what sections to print, the default being the ones from enabled collectors (they were explicitly enabled so we can assume some interest from the user). We expose an option, for now: `--show-extra-sections=[global]`.
In `EventFmt` we check if sections were not enabled, we skip it.

WDYT?
 
> For the event part, I think it has to bet loosely defined to get some flexibility. To me an event is a set of sections somehow related to each others. It might be to a packet event (aka. probe + timestamp) or to the initial configuration that is dumped by collectors (aka. timestamp only).
> 
> The difference in how those are consumed at post-processing time (is it global info? specific to an event?) is really only depending on the kind of section it is. That's up to the post-processing logic.

Apart from the comment above which I don't think changes the essence of it, I agree.

Reply:
> > Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.
> > For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.
> 
> I don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.

Right. Let me rephrase it: I don't think we should distinguish any metadata event from non-metadata ones. Everything is just an event. Some are linked to packets, some are not. But only a subset of sections is worth displaying *ever* in `collect`, `print` and `sort`. So we can actually make the distinction here (and not on the section being "metadata" or not). If a section is not meant to be displayed *ever*, let's not implement formatting logic for it.

> How about we do as with `skb` sections? In commands such as `print` and `sort`, we allow the user to define what sections to print, the default being the ones from enabled collectors (they were explicitly enabled so we can assume some interest from the user). We expose an option, for now: `--show-extra-sections=[global]`. In `EventFmt` we check if sections were not enabled, we skip it.
> 
> WDYT?

I don't think we have an example requiring this for now (introduced "md sections" are not providing direct value to the user in their own). But yes, I think it's a good way forward if we ever have some sections that might or might not be displayed, depending on what one wants. To be explicit, I think this aligns perfectly with the above proposal too, but would have no user right now.

Reply:
> > > Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.
> > > For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.
> > 
> > 
> > I don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.
> 
> Right. Let me rephrase it: I don't think we should distinguish any metadata event from non-metadata ones. Everything is just an event. Some are linked to packets, some are not. But only a subset of sections is worth displaying _ever_ in `collect`, `print` and `sort`. So we can actually make the distinction here (and not on the section being "metadata" or not). If a section is not meant to be displayed _ever_, let's not implement formatting logic for it.
> 

Sure but now all sections have formatting implemented, right? I am struggling to find a case where a section will not be printed in some way or another. Even config changes will need a way to be displayed, no?

> > How about we do as with `skb` sections? In commands such as `print` and `sort`, we allow the user to define what sections to print, the default being the ones from enabled collectors (they were explicitly enabled so we can assume some interest from the user). We expose an option, for now: `--show-extra-sections=[global]`. In `EventFmt` we check if sections were not enabled, we skip it.
> > WDYT?
> 
> I don't think we have an example requiring this for now (introduced "md sections" are not providing direct value to the user in their own). But yes, I think it's a good way forward if we ever have some sections that might or might not be displayed, depending on what one wants. To be explicit, I think this aligns perfectly with the above proposal too, but would have no user right now.

Fair enough

Reply:
> > > I don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.
> > 
> > 
> > Right. Let me rephrase it: I don't think we should distinguish any metadata event from non-metadata ones. Everything is just an event. Some are linked to packets, some are not. But only a subset of sections is worth displaying _ever_ in `collect`, `print` and `sort`. So we can actually make the distinction here (and not on the section being "metadata" or not). If a section is not meant to be displayed _ever_, let's not implement formatting logic for it.
> 
> Sure but now all sections have formatting implemented, right? I am struggling to find a case where a section will not be printed in some way or another. Even config changes will need a way to be displayed, no?

In the current form of this PR yes, but there's no reason really to implement `EventFmt` for the (currently named) `GlobalEventMd` section.

Another example I have in mind would be to dump the nft config in an event (to match at post-processing time the exact rule being hit) and here again I don't see a reason to display this in the events. The information might be used to augment what we display in some way, but the section itself has no reason to be displayed. I'm thinking this should actually apply to many sections of that kind.

Then of course we might see some sections where we'd like to print something; for those we can implement a way to select what to print.